### ISSUE TITLE
Native headerbars

### ISSUE THREAD

**Issue by 1j01** [electron/electron#11907]
Client-Side Decoration (CSD) support, for native headerbars

![eos files headerbar](https://user-images.githubusercontent.com/3630663/36128754-0773fb80-1032-11e8-96dd-c1b83a8af226.png)

* Electron version: 1.8.2 is currently the latest
* Operating system: Linux

From https://wiki.gnome.org/Initiatives/CSD:

>## Electron
>Many Electron apps should be low-hanging fruit, because they already use CSD on macOS (e.g. Slack). 
> This likely means very little work would be required on the application side to make this happen on GNU/Linux as well.
> 
> However, since some additional complexity is required in order to get the proper GNOME CSD behavior (respecting gsettings etc.), some of this should probably be supported by Electron directly. That would be easier for Electron app developers and make for more consistent behavior.
> 
> We should first work with upstream Electron to add GNOME-style CSD support at a platform level. Once that is in place, we can start approaching individual app developers.
> 

Related: https://github.com/atom/atom/issues/2215, https://github.com/electron/electron/issues/2927#issuecomment-325879914

**Comment by bugaevc** [comment]
Also related: https://github.com/atom/atom/issues/13249, https://github.com/atom/atom/issues/4599.

For Atom's titlebar implementation for macOS see https://github.com/atom/atom/pull/13616

**Comment by felixrieseberg** [comment]
Solid idea. The usual open source rules apply: If you build it, developers might use it.

**Comment by TingPing** [comment]
Dumb question but this seems to work fine on Linux: https://github.com/electron/electron/blob/master/docs/api/frameless-window.md

So what is actually needed for applications to use CSD? What method is commonly used to enable it?

**Comment by ZanderBrown** [comment]
Also relevant https://github.com/Microsoft/vscode/issues/43154

**Comment by corebots** [comment]
Please add CSD support for Electron apps on GNOME. There are so many nice Electron apps and they all still use the old titlebar :( Imagine the bomb exploding when Electron apps suddenly look like native GNOME ones :)

**Comment by MLDamazigh** [comment]
Please add CSD support on Linux

**Comment by TheGreatHalva** [comment]
Now that KDE supports headerbars too, there's even more reason to do it.

**Comment by TingPing** [comment]
Visual Studio Code has been using CSD for years on Linux so its clearly possible. It just requires the application manually opt-in for some reason.

**Comment by teohhanhui** [comment]
> Visual Studio Code has been using CSD for like a year on Linux

That's a custom implementation that doesn't respect the native look and feel.

**Comment by TingPing** [comment]
Well... yeah, thats what CSD is. The client draws whatever it wants.

**Comment by teohhanhui** [comment]
@TingPing In case you've missed it, this issue explicitly mentions **native** headerbars.

**Comment by Feichtmeier** [comment]
Uhm... No, not even in Chrome and Firefox they are native. They try to sick up as much of the OS as possible but it's never native because it's a different toolkit. So what vscode and chrome do or Firefox is all you get and it's perfectly fine

**Comment by teohhanhui** [comment]
@Feichtmeier I don't see why we're arguing over semantics here. As long as it respects the native look and feel, the user wouldn't care how it was implemented (`GtkHeaderBar` or otherwise). That's what this issue is about.

**Comment by ZanderBrown** [comment]
I _think_ the point here is that people want "native" window controls not necessarily that they want GtkHeaderBar

If people actually want GtkHeaderBar (with embedded GtkButton et al) then clearly electron isn't right for them

However some way to respect [gtk-decoration-layout](https://developer.gnome.org/gtk3/unstable/GtkSettings.html#GtkSettings--gtk-decoration-layout), ideally with gtk rendering it, would be nice to have (as on macOS with titleBarStyle/trafficLightPosition)

Of course with the differing number/position (or even size) of controls this would be interesting for apps to handle but not impossible

It's true vscode does have CSD on linux-based systems, but Windows 10 [min][max][close] controls sure look out of place on GNOME where we just have [close]

**Comment by MLDamazigh** [comment]
There are some people who prefer HeaderBars (and i am one of them) and others not, so i think the best solution is to make an option to enable or disable HeaderBars like did developers of an application called "Drawing" (https://github.com/maoschanz/drawing)
an exemple bellow of Telegram application mockup for illustration : 
HeaderBar disabled : 
![image](https://user-images.githubusercontent.com/15183134/79603058-85445180-80e3-11ea-9d35-bb5e0823a5d0.png)

HeaderBar enabled : 
![image](https://user-images.githubusercontent.com/15183134/79603107-9db46c00-80e3-11ea-8022-f5c239b25b82.png)

**Comment by ZanderBrown** [comment]
And we're back to the old mistake: That isn't a headerbar

It's CSD, and electron already lets you do what those screenshots show

In fact your "HeaderBar disabled" shot actually contains a GtkHeaderBar

What's missing is the ability to use native controls (as on macOS) and show the WM menu (as when you right click on a titlebar)

**Comment by Feichtmeier** [comment]
I would prefer CSD to a toolkit headerbar on top of a different toolkit window. what's the problem with CSD @ZanderBrown ?

**Comment by MLDamazigh** [comment]
> And we're back to the old mistake: That isn't a headerbar
> 
> It's CSD, and electron already lets you do what those screenshots show
> 
> In fact your "HeaderBar disabled" shot actually contains a GtkHeaderBar
> 
> What's missing is the ability to use native controls (as on macOS) and show the WM menu (as when you right click on a titlebar)

thanks for clarifications, i understand now, i hope that they will implement a native controls soon.

**Comment by ZanderBrown** [comment]
@Feichtmeier nothing wrong with CSD

**Comment by raelgc** [comment]
> And we're back to the old mistake: That isn't a headerbar

@ZanderBrown The component is called [HeaderBar](https://developer.gnome.org/hig/stable/header-bars.html.en) on GTK3.

QT has a movement to implement it too: [here](https://www.qt.io/blog/custom-window-decorations) and [here](https://bugreports.qt.io/browse/QTBUG-63969).

**Comment by ZanderBrown** [comment]
Correct, Gtk has GtkHeaderBar  for implementing CSD in Gtk apps

This
![Screenshot from 2020-05-03 11-29-54](https://user-images.githubusercontent.com/17074021/80911933-7095c980-8d31-11ea-91c3-ac9282b2a5dc.png)
is not a GtkHeaderBar

**Comment by raelgc** [comment]
Well, this is the default gnome file manager (Nautilus) using GtkHeaderBar:

![image](https://user-images.githubusercontent.com/263302/80914586-c6af4080-8d22-11ea-8e88-652d72a0bdf9.png)

This the file editor (Gedit):

![image](https://user-images.githubusercontent.com/263302/80914610-ee9ea400-8d22-11ea-9c5c-f142f9b607d9.png)

**Comment by ZanderBrown** [comment]
And?

**Comment by haecker-felix** [comment]
@raelgc 
![image](https://user-images.githubusercontent.com/6001248/80914862-b2813880-8d4e-11ea-9c1f-27226a2d876b.png)
This looks like a `Gtk.Headerbar`. But technically it's not a `Gtk.HeaderBar`. Telegram doesn't use GTK as toolkit. It's something else, which tries to mimic the appearance of a `Gtk.HeaderBar`

That's the point @ZanderBrown tries to make clear. 

Nautilus and Gedit are native GTK apps, so they use a real `Gtk.HeaderBar` and not something else. Telegram doesn't use GTK as toolkit, so it's not a `Gtk.HeaderBar`

**Comment by Exalm** [comment]
`GtkHeaderBar` does not exist outside of GTK. It does not and cannot exist in Qt or Electron.

You can make similar enough widgets that would look like `GtkHeaderBar`, it still won't be `GtkHeaderBar`. In fact even in GTK you can construct something that looks exactly like a headerbar without using `GtkHeaderBar`.

So "native" makes no sense here really. At best you can get a headerbar at the top of the window that won't integrate with content, like here:

![Screenshot from 2020-05-03 18-02-24](https://user-images.githubusercontent.com/1510457/80914925-5d3d2b00-8d3e-11ea-817d-758c57816ecf.png)

Yes, that's a `GtkHeaderBar`. It is totally native :) But it's probably not what you have in mind. You have in mind the design pattern, not the widget itself.

What you can do, however, is to read system decoration layout and draw similar looking buttons in the same positions + rounded corners and a window shadow.

The Telegram mockup above is an example of how that would look like. It's _not_ a `GtkHeaderBar`, but it doesn't have to be one. Having `GtkHeaderBar` in Electron is like having `GtkButton` in Electron, it simply makes no sense.

**Comment by Exalm** [comment]
Oh also, `GtkHeaderBar` is only the strip at the top of the window. It doesn't even do window dragging and right click menu by itself, that's provided by `GtkWindow`, as well as shadows and invisible border for resizing.

**Comment by Exalm** [comment]
Also note, GTK4 will have [`GtkWindowControls`](https://gnome.pages.gitlab.gnome.org/gtk/gtk/GtkWindowControls.html) widget that is basically `GtkHeaderBar`'s buttons separately. If Electron is using `GtkWindow`, it can put a `GtkOverlay` into it, with Chromium view and buttons on top of it, and provide an API for controlling button offsets and getting their dimensions.

Note however, that the buttons cannot be styled by the web view in this case. If the close button is a black ❌️, it will remain black even if the app draws dark background. GTK has API for controlling this (dark variant + the buttons can be styled via css like everything else), but Electron would have to have API for that, so probably still easier to reimplement buttons so that they are a part of the web view.

**Comment by raelgc** [comment]
@Exalm I just pasted links of QT working on something similar.

And, GTKHeaderBar DOES allow dragging, resizing, borders, right click providing similar items from titlebar and shadow (if the Gnome theme is using shadow).

**Comment by Exalm** [comment]
Not really. Dragging and right click (and double click and middle click) is not `GtkHeaderBar`, that's `GtkWindow`'s titlebar (`gtk_window_set_titlebar()`). If you place another widget there, it will still be draggable. If you place `GtkHeaderBar` inside the window, it will _not_ be draggable. :)

See https://blogs.gnome.org/alexm/2020/04/12/on-windows-and-titlebars/ for more details.

Resizing, borders and shadow similarly have nothing to do with `GtkHeaderBar`, please do more research before making claims. :)

**Comment by raelgc** [comment]
@Exalm Ok, so you found that technically is HeaderBar calling `gtk_window_set_titlebar()` to act as a titlebar. Great. But for the end user, it's the same experience.

**Comment by ZanderBrown** [comment]
From the end users perspective it's just a titlebar, who knows what's drawing it where handling what events

**Comment by Exalm** [comment]
Headerbar isn't calling anything, it's you as an app developer who puts in there. It doesn't have to be `GtkHeaderBar`.

Electron could easily be using something like `HdyWindow` from my blog post + `GtkWindowControls` in GTK4 to implement something like this, as detailed above. But not `GtkHeaderBar`, it makes no sense here.

**Comment by the-spyke** [comment]
I think "who calls who in GTK" is irrelevant. The point of this issue is "we need a way for elector app to look more like native GNOME apps". Currently, this means to replace excess semi-empty bar with app's name by a "taller" bar with useful controls and respect window buttons position (left/right) and bar's color. [GNOME blog](https://blogs.gnome.org/alexm/2020/04/12/on-windows-and-titlebars/) describes how difficult CSDs may be even in native apps.

In an Electron app you may hide the bar and use some js/css UI library for rendering GTK-like controls, but some help from Electron would be appreciated:

- providing info about current GTK config: titlebar buttons placement (left/right), theme colors.
- pointing Electron on a `div` to use it's area for native drag handle and summoning window menu.
- window borders?
- etc.

I think such solution will work for all platforms and have not issues like split headerbars (from the blog post).

**Comment by Exalm** [comment]
Right, that's a workable approach. I don't think theme colors are very useful (for example, look at Slack on [Windows](https://a.slack-edge.com/a084c/marketing/img/downloads/screenshots/windows/windows-hero-desktop.png) and [macOS](https://a.slack-edge.com/d6bad/marketing/img/downloads/screenshots/mac/mac-hero-desktop.png), so really it's about positioning controls and providing a "titlebar area".

Borders can be handled by GTK. Window corners are a thornier issue, but if you're using something like `HdyWindow`, that's automatic as well.

**Comment by ZanderBrown** [comment]
As mentioned electron already supports this on macOS, unfortunately the existing API won't work will with gtk's split controls

Not convinced about peeking at gtk colours, generally your app should just look like your app

**Comment by teohhanhui** [comment]
Theme colours are very useful, even without CSD. Take vscode for example, as a user I'd very much like vscode to respect my Gtk theme, not use whichever colours the developer see fit.

**Comment by Exalm** [comment]
@the-spyke and the funniest thing? Those colors aren't public API. _Window_ color is. _Headerbar_ color isn't.

**Comment by the-spyke** [comment]
@Exalm GTK uses CSS for styles, so there should be a way get them (I hope ^^).

**Comment by Exalm** [comment]
Not really. Especially because Yaru is using SCSS, so those colors are hardcoded.

**Comment by ZanderBrown** [comment]
(if only there was a gtk dev we could ask)

**Comment by Exalm** [comment]
A lengthier explanation:

GTK does _not_ support CSS variables. It does have a color-only "replacement": named colors. Adwaita exports a specific set of named colors, like `@theme_bg_color` for window background, `@theme_fg_color` for window text etc. Adwaita also has headerbar with similar color to the window itself, so there's no headerbar background specifically.

Now, while those aren't really an API, and there are no stability guarantees, they are kinda there and everything uses them regardless. Because of that, themes tend to export those specific colors as well because otherwise everything breaks. However, since there's no "de facto" named color for headerbar, nothing exports it.

Yaru can start exporting it. However, it would be Yaru-specific. There's no generic way to get it.

There are also some WM colors exported for Metacity (which likely include this color), but they are only there because nobody got to removing them yet, and, for example, elementary stylesheet does not export them, so still doesn't work.

Strictly speaking, there's no generic way to get any colors, but see above, those few specific named colors are kinda expected to be there (although you should still always have a fallback)

Now, this leads to some broken situations. Yaru light and dark look reasonably similar to Adwaita and so it kinda works. But the "standard" one (as well as Ambiance etc) become trainwrecks as soon as an app does something with the headerbar that the theme didn't _explicitly_ account for.

For example, this is how gnome-clocks looks like with that style:

![Screenshot from 2020-05-07 20-27-13](https://user-images.githubusercontent.com/1510457/81313320-43ba1d00-9077-11ea-8424-f795bf45424b.png)

The too-light style is hover effect. Active works simply because it's semi-transparent black and doesn't derive from the headerbar color.

Switch to dark and it works:

![Screenshot from 2020-05-07 20-28-32](https://user-images.githubusercontent.com/1510457/81313481-7106cb00-9077-11ea-8b51-367d4c5c06e3.png)

But there's more! Generally in CSS "background color" doesn't exist. Color is a part of the background, not the whole thing: there's background-image. And if the headerbar has a gradient like pretty much everywhere that's exactly the case: there's no single color you could use.

A simple experiment: open GtkInspector and type in `headerbar { background-color: red; }`. In Adwaita it will only affect the window when it's unfocused because the unfocused window is using solid color as the bg. Focused window uses a gradient, so it will do nothing.

The only reliable way to draw an element that looks like a GTK widget is to ask GTK to do it. It is possible, it's called foreign drawing and e.g. that's exactly what Firefox and Chrome do. But at that point it's:

1. Not cross-platform
2. Requires serious modifications in Chromium

So again unworkable. So really don't bother about the colors: it's not going to look native anyway.

And don't try to point at Firefox, like it does that while it renders the whole window via its engine. That integration is _deep_ in Gecko (`-moz-appearance` css property, the thing that's normally used for native-looking html forms uses foreign drawing on Linux there), and it's not cross-platform and Firefox has _vastly_ different CSS for every platform for this reason.

**Comment by the-spyke** [comment]
@Exalm Thanks for explanation. So, I'll take it without colors. In the worst scenario I can add an option to tweak the color inside my app, so the user may match the theme.

**Comment by Exalm** [comment]
But a single color will not match anything. And I mean, if you are trying to match headerbar color, then go ahead and try to match everything else as well 😉️

Headerbar is just a small part of the window, if everything else looks nothing like other apps, this particular part isn't going to suddenly make it "native".

**Comment by rathboma** [comment]
Oh man, I feel like I'm getting into the thick of it with this thread, but there seem to be several different features people are asking for in this thread. Maybe they could be broken up by the maintainers?

1. The ability to embed window controls ( - + x ) in the app without a titlebar, much like on MacOS. (This is what I would want)
2. The ability to build a legit GTK3 header bar with CSD, like native GTK apps use
3. The ability to style a HTML header bar to match the OS, with some sort of theme pass-through.

**Comment by Exalm** [comment]
> Oh man, I feel like I'm getting into the thick of it with this thread, but there seem to be several different features people are asking for in this thread. Maybe they could be broken up by the maintainers?
> 
> 1. The ability to embed window controls ( - + x ) in the app without a titlebar, much like on MacOS. (This is what I would want)
> 2. The ability to build a legit GTK3 header bar with CSD, like native GTK apps use
> 3. The ability to style a HTML header bar to match the OS, with some sort of theme pass-through.

2 and 3 are impossible. 1 is easily possible in GTK4, but in GTK3 you would have to reimplement the controls from `GtkHeaderBar`.

**Comment by ripefig** [comment]
Any particular reason why it's impossible for GTK? On Mac OS, everything is consistent with Cocoa.

Someone managed to make QT use GTK to get real headerbars etc. https://github.com/CrimsonAS/gtkplatform. These are real gtk headerbars, with the correct context menus and so on. The correspondence is 100%, not like the obviously half-assed imitations of GTK seen in LibreOffice, Firefox and Chromium. 

Obvious visual inconsistencies are only half the problem with these imitations - the behavior is fundamentally different because you can't drag these windows from widgets like you can with every real GTK headerbar. Thus something as fundamental as window dragging becomes unpredictable across apps, in addition to drag space being greatly reduced in the copycat apps (behause you have to seek out small islands of whitespace). The decorations on some Electron CSD apps are so crowded that that repositioning the window becomes a real chore.

**Comment by Exalm** [comment]
Because you cannot have a (clickable) web view behind it. You can have a web view inside it, but that's probably not what you want....

On macOS you get native title buttons overlaid above content. I already explained that it's not the case in GTK 3, but can be done in GTK 4.

So if you do this, you will get:

1. unclickable top part of the page
2. Also square window corners

You can try to quickly implement this with WebKitGTK to see exactly what will happen. :)

**Comment by xcffl** [comment]
@Exalm I guess a reasonable implementation is a HTML-based headerbar mimicing GTKHeaderBar?

**Comment by Exalm** [comment]
Well, if you can get it to be draggable and have context menu/other actions...

**Comment by Feichtmeier** [comment]
@Exalm if electron is capable of providing a window with close/min/max buttons inside the window without a window manager titlebar on top, what is the visual advantage of "painting" a gtk headerbar into the window if the rest of the window still looks like it looks. For example in vscode this could look pretty weird and personally I would prefer the current custom window, without any gtk'ish thing on top of it

**Comment by Exalm** [comment]
No advantage. Really you just need to be very careful to handle dragging and actions correctly + in GTK3 you invariably lose round corners with this.

**Comment by Feichtmeier** [comment]
Sorry for this noobish question and maybe you even answered this already in this issue, but you mean "more" than "only" repainting min/max/close, like for example chrome and firefox can already do currently?

**Comment by Exalm** [comment]
Notice how neither has proper context menu and both have broken window corners. :)

**Comment by Feichtmeier** [comment]
Firefox at least, fixed their window corner issue
![grafik](https://user-images.githubusercontent.com/15329494/101774355-463d9300-3aee-11eb-987b-a2eba0a28d5d.png)

**Comment by Exalm** [comment]
Great. But that's the header bar, if you put a web view there, it won't just have corners like that. (in GTK 3, in 4 it will because the window clips it)

And so if you do this stuff in HTML, it will be the same problem.

**Comment by ripefig** [comment]
Thanks @Exalm. So gtk4 will allow you to overlay native buttons but native decorations remain out of reach (even with gtk4), is that correct? If the limitation is with gtk, then gtk developers would have to resolve it somehow. Expecting other toolkits to carefully "imitate" gtk and then expecting developers to implement this imitation sounds like a complete non-starter. It's high effort with little reward. The imitation will never be credible and the number of users who will benefit will be tiny.

Re firefox: yeah they have fixed the round corners bug ... after about three years of grappling with this issue, and despite already having a toolkit that could use gtk drawing routines. After all that, it's still nothing like a native gtk headerbar, which becomes apparent the moment you try to use it.

**Comment by Exalm** [comment]
Well. GTK4 has GtkWindowHandle (in GTK3 you have it too, but you'd have to depend on libhandy), so you can do the following:

GtkWindow
  GtkOverlay
    web view
    GtkWindowHandle
    GtkWindowControls

I _don't_ know if you will be able to click the web view underneath window handle, that needs checking. But at least you can do it, in GTK 3 you simply cannot extend the webview under the header bar, they are completely separate areas that don't intersect, unless you once again depend on libhandy and use HdyWindow/HdyWindowHandle. :)

**Comment by ripefig** [comment]
Sorry if this is a stupid question, but is there some rational way to pass commands from gtk widgets (just rudimentary buttons and menus) to the underlying application without requiring passthrough clicks?

**Comment by Exalm** [comment]
Passthrough is kinda implicit in GTK, but I don't know for sure if specifically window handle would block it.

**Comment by Exalm** [comment]
A quick proof of concept made of GTK4 WebKit minibrowser:

![Screenshot from 2020-12-13 03-51-01](https://user-images.githubusercontent.com/1510457/101998605-9f781300-3cf6-11eb-933b-f22674c1a23f.png)

The top bar is draggable, has context menu etc. However, that area of the web page does _not_ pass any events to the web page, i.e. cannot be clicked other than for window actions, cannot be scrolled etc.

**Comment by memeplex** [comment]
Given that this is almost sine qua non for Wayland, I have opened an FR https://github.com/electron/electron/issues/27522. Only later I found out this issue which AFAICS has been disconnected from recent developments in Electron towards Wayland support. I don't know which is the best way to proceed: leaving my Wayland specific FR as part of the Wayland support initiative, leaving this more general one open, leaving both open.

**Comment by memeplex** [comment]
I'm interested in exploring the approach suggested by @Exalm here https://github.com/electron/electron/issues/11907#issuecomment-742513770.

Some questions though:

* If this were a plausible solution, would it imply full migration of Electron to GTK4?
* Would that be too hard or even acceptable?

Regarding the second question, for me the relevant time horizon is tied to Wayland support, so it's clearly not next month but, at best, next year (I'm an optimist). In that light, GTK4 seems less unreasonable.

**Comment by Exalm** [comment]
Yes, that's GTK4.

Too hard in what sense? It's just a bit of widgetry. Of course you'd also need API to control it, but it already works this way on macOS, so...

**Comment by memeplex** [comment]
> Too hard in what sense? It's just a bit of widgetry. Of course you'd also need API to control it, but it already works this way on macOS, so..

I meant porting Electron to GTK4, not GTK4 per se. I don't know how much of GTK is there in Electron, I assume it's just a scaffolding bit but it's all guessing at this point for me.

Ah, and also:

* Being GTK4 controls, would this be supported out of the box by Plasma GTK CSD integration? Anyone knows?

**Comment by memeplex** [comment]
@Exalm `GtkWindowHandle` doesn't pass through clicks, but I managed to get it somehow working with a custom gesture controller:

```python
import gi
gi.require_version("Gtk", "4.0")
from gi.repository import Gtk


def activate(app):
    win = Gtk.ApplicationWindow(application=app)
    button = Gtk.Button(label="Click me")
    button.connect("clicked", lambda *_: print("Clicked"))
    ctrls = Gtk.WindowControls(side=Gtk.PackType.END)
    ctrls.set_halign(Gtk.Align.END)
    handle = Gtk.WindowHandle()
    handle.set_valign(Gtk.Align.START)
    handle.set_child(ctrls)
    gesture = Gtk.GestureClick()
    gesture.connect("released", lambda *_: button.emit("clicked"))
    handle.add_controller(gesture)
    over = Gtk.Overlay()
    over.add_overlay(button)
    over.add_overlay(handle)
    win.set_child(over)
    win.show()

app = Gtk.Application()
app.connect("activate", activate)
app.run()
```

Here the main widget is a button, not a webview, but I think that's good enough for a POC. Even though I think it's not that bad, I sort of dislike the idea that every button release is passed to the underlying button as a click, I would like to differentiate between single and double clicks, but I don't know of a way to separate them other than waiting for a fraction of second to the second click (n_press = 2).

**Comment by memeplex** [comment]
Apart from how to do this GTK side, it seems to me that it would be hard to implement Electron side. AFAICS there is no GTK window to which simply set the overlay as content, but windows come from deep into the aura/ozone layers and have nothing GTK in them. For the MacOS case the entire `NativeWindow` was reimplemented, look at the [header file](
https://github.com/electron/electron/blob/8bf66f8974c6dc0aecdc22e4ddce2881e1836840/shell/browser/native_window_mac.h) to get an idea about how many methods were overridden, then a [WindowButtonsView](https://github.com/electron/electron/blob/8bf66f8974c6dc0aecdc22e4ddce2881e1836840/shell/browser/ui/cocoa/window_buttons_view.mm) was introduced into the window. I might be plain wrong, but I suspect something like that should be done for Linux in order to introduce the overlay, handle and buttons.

Could someone more savvy take a look at this?

**Comment by stalkerg** [comment]
What about KDE/Plasma integration? I mean I hope the Electron will not depend on GTK much.

**Comment by rohmishra** [comment]
@memeplex unfortunately im not to familiar with how electron windows work but you seem to be right. After spending a few hours in there trying to understand it, it seems that to have native headerbars from gtk, it would need huge amount of rework to get those gtk components on screen. And given that most apps dont even bother having client side decorations on linux (slack, discord, etc all use system provided bars, vscode gives an option but it is rather basic and doesnt follow system convention.) it seems unlikely anyone would bother implementing it for gtk/linux. The next best thing would be to allow apps to read the window button configuration (position, options, actual button assets) which would allow apps to use their CSD and still follow somewhat follow the system convention. This is an OK option since usability wise, we are matching the button layout and gnome HIG is open to apps using their style as long as they follow some basic conventions.

**Comment by RaisinTen** [comment]
@1j01 https://github.com/electron/electron/pull/29618 has landed already. Can we close this issue now?

**Comment by harmathy** [comment]
@RaisinTen The problem is, that it is not enabled by default and utterly broken (https://github.com/electron/electron/issues/34820)

**Comment by RaisinTen** [comment]
@harmathy I'm guessing it wasn't enabled by default because it's experimental and has the potential to be broken? To my understanding, this issue is just about having the feature in Electron and not about making sure that it's bug-free or enabled by default.

Since we have an open issue tracking the bug you mentioned already, what do you think about continuing the discussion there / creating newer issues if more bugs are found? FWIW, the PR author mentioned that they'll take a look at the issue you shared - https://github.com/electron/electron/issues/34820#issuecomment-1175651974.

**Comment by RaisinTen** [comment]
I see there are few PRs up already, attempting to fix that bug. Closing this issue for now, thank you all for the feature request!
